src/axom/quest/MarchingCubesAlgo.cpp

```diff
@@ -0,0 +1,614 @@
+  const int coordSp = isInterleaved ? _ndim : 1;
+  const double* xPtr = coordValues["x"].as_double_ptr();
+  const double* yPtr = _ndim >= 2 ? coordValues["y"].as_double_ptr() : nullptr;
+  const double* zPtr = _ndim >= 3 ? coordValues["z"].as_double_ptr() : nullptr;
+
+  auto& fcnValues = _dom->fetch_existing(_fcnPath + "/values");
+  const double* fcnPtr = fcnValues.as_double_ptr();
+
+  const int* maskPtr = nullptr;
+  if(!_maskPath.empty())
+  {
+    auto& maskValues = _dom->fetch_existing(_maskPath + "/values");
+    maskPtr = maskValues.as_int_ptr();
+  }
+
+  if(!_cellIdField.empty() &&
+     !_surfaceMesh->hasField(_cellIdField, axom::mint::CELL_CENTERED))
+  {
+    _surfaceMesh->createField<int>(_cellIdField, axom::mint::CELL_CENTERED);
+  }
+
+  _contourVal = contourVal;
+
+  if(_ndim == 2)
+  {
+    /*
+      For now, assume zero offsets and zero ghost width.
+      Eventually, we'll have to support index offsets to
+      handle data with ghosts.
+
+      By using ArrayView, we're assuming row-major layout.  To support
+      column-major layout as well, we have to extend ArrayView to
+      compute column-major strides.  We should also use an iteration
+      scheme that's efficient for both layouts.
+    */
+    const axom::StackArray<axom::IndexType, 2> cShape {_cShape[0], _cShape[1]};
+    const axom::StackArray<axom::IndexType, 2> pShape {1 + _cShape[0],
+                                                       1 + _cShape[1]};
+    axom::ArrayView<const double, 2> xView(xPtr, pShape, coordSp);
+    axom::ArrayView<const double, 2> yView(yPtr, pShape, coordSp);
+    axom::ArrayView<const double, 2> fcnView(fcnPtr, pShape);
+    axom::ArrayView<const int, 2> maskView(maskPtr, cShape);
+
+    // Write as regular nested loops.
+    for(int i = 0; i < cShape[0]; ++i)
+    {
+      for(int j = 0; j < cShape[1]; ++j)
+      {
+        const bool skipZone = maskPtr && bool(maskView(i, j));
+        if(!skipZone)
+        {
+          double vfs[4];
+          double xx[4];
+          double yy[4];
+
+          vfs[0] = fcnView(i + 1, j);
+          vfs[1] = fcnView(i + 1, j + 1);
+          vfs[2] = fcnView(i, j + 1);
+          vfs[3] = fcnView(i, j);
+
+          xx[0] = xView(i + 1, j);
+          xx[1] = xView(i + 1, j + 1);
+          xx[2] = xView(i, j + 1);
+          xx[3] = xView(i, j);
+
+          yy[0] = yView(i + 1, j);
+          yy[1] = yView(i + 1, j + 1);
+          yy[2] = yView(i, j + 1);
+          yy[3] = yView(i, j);
+
+          auto nPrev = _surfaceMesh->getNumberOfCells();
+          this->contourCell2D(xx, yy, vfs);
+          auto nNew = _surfaceMesh->getNumberOfCells();
+
+          if(nNew > nPrev && !_cellIdField.empty())
+          {
+            int zoneIdx = i + j * cShape[0];  // TODO: Fix for ghost layer size.
+            auto* cellIdPtr =
+              _surfaceMesh->getFieldPtr<int>(_cellIdField,
+                                             axom::mint::CELL_CENTERED);
```

Comment:
Can this pointer extraction get hoisted out of the doubly-nested loop?

Reply:
It has to be in the loop.  Adding cells to the Mint mesh may cause a data reallocation.